File: app/src/test/java/com/papi/nova/utils/CacheHelperTest.java
package com.papi.nova.utils;

import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

public class CacheHelperTest {
    @Rule
    public TemporaryFolder temporaryFolder = new TemporaryFolder();

    @Test
    public void cacheStreamsAllowSafeComponents() throws IOException {
        File root = temporaryFolder.newFolder("cache");

        try (OutputStream outputStream = CacheHelper.openCacheFileForOutput(
                root, "boxart", "host-uuid", "123.png")) {
            CacheHelper.writeStringToOutputStream(outputStream, "ok");
        }

        try (InputStream inputStream = CacheHelper.openCacheFileForInput(
                root, "boxart", "host-uuid", "123.png")) {
            assertEquals("ok", CacheHelper.readInputStreamToString(inputStream));
        }
    }

    @Test
    public void cacheStreamsRejectParentTraversalComponents() throws IOException {
        File root = temporaryFolder.newFolder("cache");

        try {
            CacheHelper.openCacheFileForOutput(root, "boxart", "..", "escape.png");
            fail("Expected parent traversal component to be rejected");
        } catch (IOException expected) {
            // Expected.
        }
    }

    @Test
    public void cacheStreamsRejectEmbeddedSeparators() throws IOException {
        File root = temporaryFolder.newFolder("cache");

        try {
            CacheHelper.openCacheFileForInput(root, "boxart/escape.png");
            fail("Expected embedded separator to be rejected");
        } catch (IOException expected) {
            // Expected.
        }
    }
}

File: app/src/test/java/com/papi/nova/utils/CacheHelperTest.kt
package com.papi.nova.utils

import java.io.File
import java.io.IOException
import java.nio.file.Files
import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.junit.Assume.assumeNoException
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder

class CacheHelperTest {
    @get:Rule
    val temporaryFolder = TemporaryFolder()

    @Test
    fun cacheStreamsAllowSafeComponents() {
        val root = temporaryFolder.newFolder("cache")

        CacheHelper.openCacheFileForOutput(root, "boxart", "host-uuid", "123.png").use { outputStream ->
            CacheHelper.writeStringToOutputStream(outputStream, "ok")
        }

        CacheHelper.openCacheFileForInput(root, "boxart", "host-uuid", "123.png").use { inputStream ->
            assertEquals("ok", CacheHelper.readInputStreamToString(inputStream))
        }
    }

    @Test
    fun cacheStreamsRejectParentTraversalComponents() {
        val root = temporaryFolder.newFolder("cache")

        try {
            CacheHelper.openCacheFileForOutput(root, "boxart", "..", "escape.png")
            fail("Expected parent traversal component to be rejected")
        } catch (expected: IOException) {
            // Expected.
        }
    }

    @Test
    fun cacheStreamsRejectEmbeddedSeparators() {
        val root = temporaryFolder.newFolder("cache")

        try {
            CacheHelper.openCacheFileForInput(root, "boxart/escape.png")
            fail("Expected embedded separator to be rejected")
        } catch (expected: IOException) {
            // Expected.
        }
    }

    @Test
    fun cacheStreamsRejectSymlinkEscapes() {
        val root = temporaryFolder.newFolder("cache")
        val outside = temporaryFolder.newFolder("outside")
        val secret = File(outside, "secret.txt").apply {
            writeText("secret")
        }
        val link = File(root, "link")
        try {
            Files.createSymbolicLink(link.toPath(), outside.toPath())
        } catch (e: UnsupportedOperationException) {
            assumeNoException(e)
        } catch (e: IOException) {
            assumeNoException(e)
        }

        try {
            CacheHelper.openCacheFileForInput(root, "link", secret.name)
            fail("Expected cache stream to reject symlink escape")
        } catch (expected: IOException) {
            // Expected.
        }
    }
}
